### ISSUE TITLE
JSON argument source fails with LocalDate property deserialization

### ISSUE THREAD

**Issue by virtual-machinist** [junit-pioneer/junit-pioneer#624]
For some odd reason I get an exception while deserializing `java.time.LocalDate` properties:

```
Failed to convert to type class com.foo.bar.MyClass
java.io.UncheckedIOException: Failed to convert to type class com.foo.bar.MyClass
	at org.junitpioneer.jupiter.json.JacksonNode.toType(JacksonNode.java:51)
	at org.junitpioneer.jupiter.json.JacksonNode.value(JacksonNode.java:85)
...	
Caused by: com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Java 8 date/time type `java.time.LocalDate` not supported by default: add Module "com.fasterxml.jackson.datatype:jackson-datatype-jsr310" to enable handling
 at [Source: UNKNOWN; byte offset: #UNKNOWN] (through reference chain: com.foo.bar.MyClass["someDate"])
```

The JSR-310 Jackson datatype dependency is in the classpath, but upon further inspection the ObjectMapper instance in `org.junitpioneer.jupiter.json.JacksonJsonConverter` has no registered modules.

Any way to fix this? Or better yet provide a way to customize the used ObjectMapper instance? Right now I'm working around this by using reflection to manually register the module.

Using Gradle 7.1.1 on OpenJDK 11 with classpath.

**Comment by beatngu13** [comment]
Hi @virtual-machinist,

thx for reaching out – we will look into this.

**Comment by robtimus** [comment]
You'll need to register the JavaTimeModule (see https://github.com/FasterXML/jackson-modules-java8).

**Comment by robtimus** [comment]
`ObjectMapper` instances don't register modules that are available on the class path automatically. There is a method though, [findAndRegisterModules()](https://fasterxml.github.io/jackson-databind/javadoc/2.7/com/fasterxml/jackson/databind/ObjectMapper.html#findAndRegisterModules()).

**Comment by virtual-machinist** [comment]
That's actually the first thing I've tried before eventually resorting to `objectMapper.registerModule(new JavaTimeModule())`. For some reason no modules were found. I'll recheck this next week. I'm betting either `ServiceLoader` broken and finds no modules in classpath or me looking the wrong way while debugging :see_no_evil: .

Anyhow. That's beside the point. At the moment I cannot access the `ObjectMapper` instance used by the argument source without a heap of ugly reflection to do even something trivial as [findAndRegisterModules()](https://fasterxml.github.io/jackson-databind/javadoc/2.7/com/fasterxml/jackson/databind/ObjectMapper.html#findAndRegisterModules()). Is this something worth improving? Or at least it'd be nice to switch the `JsonConverter` implementation in `JsonConverterProvider`.

**Comment by robtimus** [comment]
Aren't you simply missing a dependency? `JavaTimeModule` isn't in the default databind dependency, but instead this one: https://mvnrepository.com/artifact/com.fasterxml.jackson.datatype/jackson-datatype-jsr310

**Comment by virtual-machinist** [comment]
Yes, I understand that it's not in the default databind dependency. The error message in the issue post already suggests this. :wink: 

And yes, I have it in classpath. To be clear the tests that use JSON argument source are part of a larger test suite using among many other things [Spring Boot JSON starter](https://mvnrepository.com/artifact/org.springframework.boot/spring-boot-starter-json), that adds all the needed dependencies.

**But even if I am missing this dependency** like you suggest, how does knowing that (i.e. I have to add it, register the module, etc.) help when I have to use a ton of reflection to get hold of the `ObjectMapper` instances used by this JUnit extension? The thing I'm talking about -

```
// non-public class, already a bummer :/
class JacksonJsonConverter implements JsonConverter {

        // need to get this singleton, either directly or using the package private getConverter() method
	private static final JacksonJsonConverter INSTANCE = new JacksonJsonConverter(new ObjectMapper());

        // both have to be changed, since one is a clone of the other with extra settings?
	private final ObjectMapper objectMapper;
	private final ObjectMapper lenientObjectMapper;

```

**Comment by robtimus** [comment]
Sorry, I thought you were a Pioneer developer. I'm not either, I am just someone who's interested.

I was trying to work on a pull request, but I added a single test and without any other changes it succeeded. That's not what I was expecting. This is the test method signature:

```
@ParameterizedTest
@JsonSource({ "{ testDate: '2022-05-11' }", "{ testDate: '1970-01-01' }" })
void javaTime(@Property("testDate") LocalDate date)
```

Can you show your test method signature? Maybe you're doing something I'm missing.

**Comment by virtual-machinist** [comment]
Sorry, not at the moment, I have no access to either the code or the tests. But there's pretty much the original exception thrown by Jackson in my original post. Try creating a DTO with this property, not referencing it through `@Property("testDate")`.

**Comment by robtimus** [comment]
I could recreate it with a POJO. Adding the `.findAndRegisterModules()` then fixes the test. If I then remove the JSR310 dependency the test fails again.

I'll create a pull request shortly.

**Comment by robtimus** [comment]
See #629

**Comment by Bukama** [comment]
@robtimus and @virtual-machinist Thank your for adding the discussion and the PR!

**Comment by nipafx** [comment]
I have a few thoughts on how to handle this issue, but they are rather general and potentially tainted by my lack of deep understanding of Jackson (and other JSON parsers for that matter). Let me know, when I'm being stupid. 😉 

Ideally, we'd have a configuration mechanism that gives users full control over the JSON parser:

* which parser to use (if there are several)
* how to configure it

That mechanism would basically require the user to provide a fully configured parser.

From there I can see three steps to make things simpler:

1. Pioneer picks the parser and the user just configures it
2. Pioneer picks the parser and configures it based on some presets (e.g. "no modules", "selected modules", "all modules")
3. Pioneer picks one of those configurations as default

Does that make sense? If so, how can we create a mechanism that gives us the flexibility to make all those options simple-ish to use?

**Comment by nipafx** [comment]
We also need to consider the common case that users will only want to define very few or even just one parser for their entire test suite. That means, we can't just rely on annotations on a method in the test class.

**Comment by filiphr** [comment]
I was waiting for the stream to discuss all this options, but I missed it in the end.

>Ideally, we'd have a configuration mechanism that gives users full control over the JSON parser:
>
> * which parser to use (if there are several)
> * how to configure it

I was already having these things in mind when I was working on this feature.

What I had envisioned back then were the following options:

* `@JsonParser("jackson")` or `@JsonParser("gson")` that can be used on a method or a class. This has highest precedence and we can use it to let the user decide which parser to use. I don't really see the benefit of users using this.
* A configuration option that can be acquired through `ExtensionContext#getConfigurationParameter` e.g. `context.getConfigurationParameter("junitpioneer.json.parser")` in the `JsonConverterProvider`. The users can define `jackson`, `gson`, etc. here. This would allow the users to define the global parser in the `junit-platform.properties`. We can rely on the Mechanism from Jupiter here


The properties above are interesting when a user has multiple Json parsers available and they do not like to use the default.

The next set of options would be a user to configure a specific parser based on whether it is `jackson`, `gson` or something else. In the case of `jackson` to provide an `ObjectMapper`.

In my opinion it is enough for Pioneer to provide an already configured `ObjectMapper` with some of the well known modules, if they are on the classpath. There is no need to do "no modules", "selected modules", "all modules" etc.

As a final resort we can expose an SPI for something like `JacksonJsonConverterObjectMapperProvider` that is going to return an already configured `ObjectMapper`. The users can then do whatever they want here. This is the most flexible one.

@nipafx if you are interested in these options, we could work on them together and clarify things more easily perhaps?

-------

In https://github.com/junit-pioneer/junit-pioneer/pull/629#issuecomment-1130413330 @robtimus was asking about the dependencies. I think that it is OK for Pioneer to have `jackson-datatype-jdk8` and `jackson-datatype-jsr310` when a user does `requireCapability("org.junit-pioneer:junit-pioneer-jackson")`. If they want kotlin, the user will need to add that dependency on their own.

We have to remember that a lot of users are using Maven, which means, that they have to define which Jackson dependencies they want to have. On top of that, the combination of the default, JDK 8 module and JavaTime module is enough for 99% of the use case in my opinion. 

In addition, a user usually might have a dependency on some other framework, e.g. Spring Boot, Micronaut, Quarkus and have all the dependencies already available on their production path.

**Comment by robtimus** [comment]
For #629 I have some uncommitted changes that register a `Jdk8Module` and a `JavaTimeModule` if these are available, and a `KotlinModule` if it's available and so is Kotlin (check on class `kotlin.Metadata`). I can create a new PR for that.

Maven users indeed would need to add `jackson-datatype-jdk8` and `jackson-datatype-jsr310` to their dependencies, but odds are that they already have these as compile or runtime dependencies if they need them in their JSON test definitions, because they most likely need them in their main code as well.

**Comment by nipafx** [comment]
> Is there any reason why you can't just expose the object mapper and everyone that needs to configure it grabs the reference?

Of course we _can_ but we don't want to. 😉 First, it would make the class and the field/method that exposes the mapper part of our public API. This may prevent future bug fixes and new features, which is why barely any extension implementation is public. Second, the user code that configures this mapper needs to be reachable from _every test class_. If configuration only happens in one place, all tests running before or all partial test suite runs that don't include that class would run with an unconfigured mapper. The best way to implement this would be a custom annotation that then needs to be present on every test class. Doable but not great.

**Comment by nipafx** [comment]
But thanks for commenting on this issue @meredrica, it brought it to our attention again. With 2.0 about to be released, this seems like a good time to tackle this issue.

@filiphr, @robtimus: Sorry for not getting back to you - your comments either slipped by unnoticed or I forgot to reply. Thanks for your ideas and readiness to implement.

I re-read this issue and #629 and tried to come up with a minimal solution that focuses on:

* just Jackson (i.e. no other parsers)
* only global configuration (i.e. not per test (class))

I just created #704 and #705 - they would give us a relatively simple solution for the immediate issue and an escape hatch for everything more complicated. Once that's done we can tackle in-between options (e.g. preconfiguring a mapper and passing it to the user for further refinement) as well as other parsers (e.g. GSON).

If we agree on the overall approach (looking at all the JSON users in this thread 👀), I'll close this issue, so we can work on the other two.

**Comment by nipafx** [comment]
But thanks for commenting on this issue @meredrica, it brought it to our attention again. With 2.0 about to be released, this seems like a good time to tackle this issue.

@filiphr, @robtimus: Sorry for not getting back to you - your comments either slipped by unnoticed or I forgot to reply. Thanks for your ideas and readiness to implement.

I re-read this issue and #629 and tried to come up with a minimal solution that focuses on:

* just Jackson (i.e. no other parsers)
* only global configuration (i.e. not per test (class))

I just created #704 and #705 - they would give us a relatively simple solution for the immediate issue and an escape hatch for everything more complicated. Once that's done we can tackle in-between options (e.g. preconfiguring a mapper and passing it to the user for further refinement) as well as other parsers (e.g. GSON).

If we agree on the overall approach (looking at all the JSON users in this thread 👀), I'll close this issue, so we can work on the other two.

**Comment by meredrica** [comment]
I guess everybody will use the same Jackson config for all tests.

In the last 10 years I've never seen a project that actually has different setups in the same deployable. In fact, it's an annoying source of bugs to have more than one.

Jackson is also the de facto standard that's in use pretty much everywhere. GSON is catching up, but still not quite there.

And the overall situation would defiantly change to the better, right now there is no way to use GSON anyway 😜🤔

**Comment by Michael1993** [comment]
This was solved (I think) in #741?

**Comment by Bukama** [comment]
> This was solved (I think) in #741?

Yes I think so too. If anybody think it is not feel free to reach out to us please.
